### DISCUSSION TITLE
First commit pushed triggers 2 workflow runs

### DISCUSSION THREAD

**Discussion by johnstonmatt** [community/community#50356] [Actions]
Bug

### Body

The first commit to a GitHub repository with a on: push: event workflow results in that action being executed twice. The action running twice per one commit is shown plainly in the Actions view. In our particular use case this represents deploying two clusters to the cloud with Terraform, so the issue has potential to be high-impact and costly for users.

Reproduction Steps

1. Create an empty GitHub repo
2. Add a workflow that is triggered by a push event to main
3. Make your first commit by pushing up the workflow
4. Observe that the workflow is executed twice for a single commit

I've made a demo repo: https://github.com/polyseam/dupe-bug-demo

**Comment by Mohamed-SayedAlAhl** [comment]
It appears that the issue you are experiencing is related to a GitHub Actions workflow being triggered twice for a single commit. This can result in high-impact and potentially costly consequences for users, such as deploying two clusters to the cloud with Terraform.

To reproduce the issue, you have provided steps that involve creating an empty GitHub repo, adding a workflow triggered by a push event to main, and making the first commit by pushing up the workflow. You have observed that the workflow is executed twice for a single commit.

To address this issue, you may want to try the following solutions:

1. Check if you have any duplicate workflows or actions that are causing the workflow to run twice. If so, remove the duplicates to prevent the issue from occurring.

2. Ensure that your workflow file does not contain any unnecessary triggers or conditions that may be causing it to run twice. You can check this by reviewing the YAML syntax of the workflow file and making any necessary adjustments.

3. If the issue persists, you can try using the `on: push: branches: main` syntax instead of `on: push:`, which may help to prevent the workflow from running twice for a single commit.

If none of these solutions work, you may want to consider reaching out to GitHub support for further assistance.

**Comment by johndietz** [comment]
wanted to weigh in that we at kubefirst.io have also found this to be the case with our platforms that install to github. with high frequency, we create initial empty github repositories, and then push an initial commit to those repositories with repo content that includes github actions workflows designed to trigger from main. the initial push of the repo always fires 2 identical github actions pipelines off of the single commit sha pushed. for our particular workflow, one will pass and one will fail due to the git commit sha being a part of our immutable container naming convention. agree this should be a github issue.

**Comment by johnstonmatt** [comment]
@martinwoodward has [confirmed](https://twitter.com/martinwoodward/status/1637042988307161089) this issue and a fix should be coming down the pipe soon

**Reply by MissaelHRC** [reply]
Hi @johnstonmatt reading your last answer, do you know if the GitHub team has an issue opened to fix it?
I have the same issue with my team triggering a workflow after installing a GitHub App that we are developing. Thanks!

**Comment by thantzin-soe** [comment]
same here.I add a workflow to my laravel project and then push to my repository first time,then the workflow trigger twice.But in the next commit the workflow trigger once.Can someone explain me that this is a bug from github itself or something?

**Reply by johnstonmatt** [reply]
This is a bug in GitHub itself

**Comment by johnstonmatt** [comment]
This is still an issue as of today 
https://github.com/polyseam/dupe-demo/actions

@MissaelHRC 

There is a ticket:

<img width="972" alt="duplicate-workflow-exec" src="https://github.com/community/community/assets/5455542/8eb13263-0182-44a0-b01e-3927b1aa9a98">

**Reply by MissaelHRC** [reply]
@johnstonmatt  Thanks for the answer, I'll continue to follow this issue.

**Comment by bowtie-ltsa** [comment]
Here's my work-around, for whatever it is worth:

Background:
- I am using golang to create repositories.
- The 1st commit is the one created through "autoInit" by github (i.e. it contains a single oneliner README file.)
- The 2nd commit contains the "real" payload, a set of files that includes workflow.

When that 2nd commit triggers two workflow runs instead of one, it is catastrophic for my use case.

Work-around for this nasty bug:
- Ensure that the 2nd commit contains a single benign workflow, and just let the bug do its thing.
- Then it is safe to make the 3rd commit  with the "real" payload. (There is no need to wait until the buggy benign workflow finishes.)

Work-around attempts that did not work reliably:
- bug still present: Wait 2 seconds between autoInit and 2nd commit.
- bug still present: Ensure the 2nd commit is not empty but still contains *no* workflow at all, then put the real payload in 3rd commit.

If anyone is interested in my functional work-around, in poorly written golang code, I could post it... not sure how helpful it would be...

**Reply by johnstonmatt** [reply]
thanks for sharing!

we landed on using an action which "locks" out any subsequent runs from being called if there is a run in progress, then releases the lock when it is finished

https://github.com/polyseam/cndi/blob/d3f684284755fb353f1497285e626e867058f350/src/outputs/cndi-run-workflow.ts#LL28C1-L28C1

**Comment by alexpotv** [comment]
Also having this issue! Using a template repository, which contains a workflow to initialize Terraform. Upon cloning the template, the initial commit triggers two simultaneous runs of the workflow, resulting in a lock error.

**Comment by JoshMcCullough** [comment]
This is an issue for us but is not catastrophic as it may be with others. But it is marginally annoying. I'll try @bowtie-ltsa's suggestion of pushing a benign workflow first.

**Comment by Twipped** [comment]
FYI you can work around this by adding a concurrency interrupt to your first workflow job.

```yaml
jobs:
  my-job:
    concurrency:
      group: ${{ github.head_ref }}-init
      cancel-in-progress: true
    runs-on: ubuntu-latest
    steps:
```

This will interrupt the first execution before it even starts.

**Comment by gm3dmo** [comment]
Engineering advised that a fix has been shipped.

**Comment by olga-ibragimova-mckinsey** [comment]
@johnstonmatt Hi Matt)) do you have any updates regarding the ticket you've created? We face the same issue with triggering pipelines twice

**Reply by johnstonmatt** [reply]
Hi Olga, I believed that the fix was shipped, maybe there has been a regression?

**Comment by ghost** [comment]
I'm seeing my CI actions run on twice every single push to the repo with just the default config. It's almost as annoying as Dependabot being automatically enabled without my consent.  

```
on:
  push:
    branches: [main]
```
